File: JavaScript/url_utils/url_utils_test.js
/**
 * URL Utilities Test Suite
 * Comprehensive tests for all url_utils functions
 */

const assert = require('assert');
const urlUtils = require('./mod.js');

let passCount = 0;
let failCount = 0;

function test(name, fn) {
    try {
        fn();
        console.log(`✓ ${name}`);
        passCount++;
    } catch (e) {
        console.log(`✗ ${name}: ${e.message}`);
        failCount++;
    }
}

console.log('\n=== URL Utilities Test Suite ===\n');

// Test parse()
console.log('\n--- parse() ---');

test('parse basic URL', () => {
    const result = urlUtils.parse('https://example.com/path');
    assert.strictEqual(result.protocol, 'https');
    assert.strictEqual(result.hostname, 'example.com');
    assert.strictEqual(result.path, '/path');
});

test('parse URL with port', () => {
    const result = urlUtils.parse('https://example.com:8080/path');
    assert.strictEqual(result.port, '8080');
});

test('parse URL with query', () => {
    const result = urlUtils.parse('https://example.com?key=value&foo=bar');
    assert.deepStrictEqual(result.query, { key: 'value', foo: 'bar' });
});

test('parse URL with hash', () => {
    const result = urlUtils.parse('https://example.com#section');
    assert.strictEqual(result.hash, 'section');
});

test('parse URL with authentication', () => {
    const result = urlUtils.parse('https://user:pass@example.com');
    assert.strictEqual(result.username, 'user');
    assert.strictEqual(result.password, 'pass');
});

test('parse invalid URL returns null', () => {
    assert.strictEqual(urlUtils.parse(''), null);
    assert.strictEqual(urlUtils.parse(null), null);
    assert.strictEqual(urlUtils.parse(undefined), null);
    assert.strictEqual(urlUtils.parse(123), null);
});

// Test parseQuery()
console.log('\n--- parseQuery() ---');

test('parse basic query string', () => {
    const result = urlUtils.parseQuery('?key=value');
    assert.deepStrictEqual(result, { key: 'value' });
});

test('parse query string without leading ?', () => {
    const result = urlUtils.parseQuery('key=value');
    assert.deepStrictEqual(result, { key: 'value' });
});

test('parse multiple parameters', () => {
    const result = urlUtils.parseQuery('a=1&b=2&c=3');
    assert.deepStrictEqual(result, { a: '1', b: '2', c: '3' });
});

test('parse encoded values', () => {
    const result = urlUtils.parseQuery('name=John%20Doe&email=test%40example.com');
    assert.strictEqual(result.name, 'John Doe');
    assert.strictEqual(result.email, 'test@example.com');
});

test('parse duplicate keys to array', () => {
    const result = urlUtils.parseQuery('tag=a&tag=b&tag=c');
    assert.deepStrictEqual(result.tag, ['a', 'b', 'c']);
});

test('parse array notation', () => {
    const result = urlUtils.parseQuery('items[]=1&items[]=2');
    assert.deepStrictEqual(result.items, ['1', '2']);
});

test('parse empty query string', () => {
    const result = urlUtils.parseQuery('');
    assert.deepStrictEqual(result, {});
});

test('parse parameter without value', () => {
    const result = urlUtils.parseQuery('flag');
    assert.strictEqual(result.flag, '');
});

// Test buildQuery()
console.log('\n--- buildQuery() ---');

test('build basic query string', () => {
    const result = urlUtils.buildQuery({ key: 'value' });
    assert.strictEqual(result, '?key=value');
});

test('build query without prefix', () => {
    const result = urlUtils.buildQuery({ key: 'value' }, { prefix: '' });
    assert.strictEqual(result, 'key=value');
});

test('build query with multiple parameters', () => {
    const result = urlUtils.buildQuery({ a: '1', b: '2' });
    assert.strictEqual(result, '?a=1&b=2');
});

test('build query with sorted keys', () => {
    const result = urlUtils.buildQuery({ z: '1', a: '2' }, { sort: true });
    assert.strictEqual(result, '?a=2&z=1');
});

test('build query with array values', () => {
    const result = urlUtils.buildQuery({ tags: ['a', 'b'] });
    assert.ok(result.includes('tags=a'));
    assert.ok(result.includes('tags=b'));
});

test('build query skips null/undefined values', () => {
    const result = urlUtils.buildQuery({ a: '1', b: null, c: undefined });
    assert.strictEqual(result, '?a=1');
});

test('build empty object returns empty string', () => {
    assert.strictEqual(urlUtils.buildQuery({}), '');
    assert.strictEqual(urlUtils.buildQuery(null), '');
});

// Test build()
console.log('\n--- build() ---');

test('build basic URL', () => {
    const result = urlUtils.build({ protocol: 'https', hostname: 'example.com' });
    assert.strictEqual(result, 'https://example.com/');
});

test('build URL with port', () => {
    const result = urlUtils.build({ protocol: 'https', hostname: 'example.com', port: '8080' });
    assert.strictEqual(result, 'https://example.com:8080/');
});

test('build URL with path', () => {
    const result = urlUtils.build({ protocol: 'https', hostname: 'example.com', path: '/api/users' });
    assert.strictEqual(result, 'https://example.com/api/users');
});

test('build URL with query', () => {
    const result = urlUtils.build({ protocol: 'https', hostname: 'example.com', query: { key: 'value' } });
    assert.strictEqual(result, 'https://example.com/?key=value');
});

test('build URL with authentication', () => {
    const result = urlUtils.build({ protocol: 'https', hostname: 'example.com', username: 'user', password: 'pass' });
    assert.strictEqual(result, 'https://user:pass@example.com/');
});

test('build URL with hash', () => {
    const result = urlUtils.build({ protocol: 'https', hostname: 'example.com', hash: 'section' });
    assert.strictEqual(result, 'https://example.com/#section');
});

test('build throws without hostname', () => {
    assert.throws(() => urlUtils.build({}), /Hostname is required/);
});

// Test isValid()
console.log('\n--- isValid() ---');

test('valid HTTP URL', () => {
    assert.strictEqual(urlUtils.isValid('http://example.com'), true);
});

test('valid HTTPS URL', () => {
    assert.strictEqual(urlUtils.isValid('https://example.com'), true);
});

test('invalid URL returns false', () => {
    assert.strictEqual(urlUtils.isValid('not a url'), false);
    assert.strictEqual(urlUtils.isValid(''), false);
});

test('disallowed protocol returns false', () => {
    assert.strictEqual(urlUtils.isValid('ftp://example.com'), false);
});

test('allow custom protocols', () => {
    assert.strictEqual(urlUtils.isValid('ftp://example.com', { protocols: ['ftp'] }), true);
});

test('URL without protocol when requireProtocol=false', () => {
    assert.strictEqual(urlUtils.isValid('example.com/path', { requireProtocol: false }), true);
});

// Test joinPaths()
console.log('\n--- joinPaths() ---');

test('join simple paths', () => {
    assert.strictEqual(urlUtils.joinPaths('a', 'b', 'c'), 'a/b/c');
});

test('join with leading slashes', () => {
    assert.strictEqual(urlUtils.joinPaths('/a', '/b', '/c'), '/a/b/c');
});

test('join empty parts', () => {
    assert.strictEqual(urlUtils.joinPaths('a', '', 'b'), 'a/b');
});

test('join with full URL', () => {
    const result = urlUtils.joinPaths('https://example.com/', 'path');
    assert.strictEqual(result, 'https://example.com/path');
});

test('join no arguments returns root', () => {
    assert.strictEqual(urlUtils.joinPaths(), '/');
});

// Test resolve()
console.log('\n--- resolve() ---');

test('resolve relative path', () => {
    assert.strictEqual(
        urlUtils.resolve('https://example.com/page', '/api/data'),
        'https://example.com/api/data'
    );
});

test('resolve relative path without leading slash', () => {
    assert.strictEqual(
        urlUtils.resolve('https://example.com/', 'api/data'),
        'https://example.com/api/data'
    );
});

test('resolve with query parameters', () => {
    const result = urlUtils.resolve('https://example.com', '?key=value');
    assert.strictEqual(result, 'https://example.com/?key=value');
});

test('resolve with null base returns null', () => {
    assert.strictEqual(urlUtils.resolve(null, '/path'), null);
});

// Test normalize()
console.log('\n--- normalize() ---');

test('normalize removes default HTTP port', () => {
    const result = urlUtils.normalize('http://example.com:80/path');
    assert.strictEqual(result, 'http://example.com/path');
});

test('normalize removes default HTTPS port', () => {
    const result = urlUtils.normalize('https://example.com:443/path');
    assert.strictEqual(result, 'https://example.com/path');
});

test('normalize lowercases hostname', () => {
    const result = urlUtils.normalize('https://EXAMPLE.COM/path');
    assert.strictEqual(result, 'https://example.com/path');
});

test('normalize removes trailing slash', () => {
    const result = urlUtils.normalize('https://example.com/path/');
    assert.strictEqual(result, 'https://example.com/path');
});

test('normalize keeps root slash', () => {
    const result = urlUtils.normalize('https://example.com/');
    assert.strictEqual(result, 'https://example.com/');
});

test('normalize sorts query parameters', () => {
    const result = urlUtils.normalize('https://example.com?z=1&a=2');
    assert.strictEqual(result, 'https://example.com/?a=2&z=1');
});

test('normalize returns null for invalid URL', () => {
    assert.strictEqual(urlUtils.normalize('not a url'), null);
});

// Test getDomain()
console.log('\n--- getDomain() ---');

test('get domain from URL', () => {
    assert.strictEqual(urlUtils.getDomain('https://www.example.com/path'), 'www.example.com');
});

test('get domain with port', () => {
    assert.strictEqual(urlUtils.getDomain('https://example.com:8080'), 'example.com');
});

test('get domain returns null for invalid URL', () => {
    assert.strictEqual(urlUtils.getDomain('not a url'), null);
});

// Test getTLD()
console.log('\n--- getTLD() ---');

test('get TLD from URL', () => {
    assert.strictEqual(urlUtils.getTLD('https://example.com'), 'com');
});

test('get TLD from subdomain URL', () => {
    assert.strictEqual(urlUtils.getTLD('https://blog.example.co.uk'), 'uk');
});

test('get TLD returns null for IP address', () => {
    assert.strictEqual(urlUtils.getTLD('https://192.168.1.1'), null);
});

test('get TLD returns null for localhost', () => {
    assert.strictEqual(urlUtils.getTLD('https://localhost'), null);
});

// Test getSubdomain()
console.log('\n--- getSubdomain() ---');

test('get subdomain from URL', () => {
    assert.strictEqual(urlUtils.getSubdomain('https://blog.example.com'), 'blog');
});

test('get subdomain from multi-level subdomain', () => {
    assert.strictEqual(urlUtils.getSubdomain('https://api.v1.example.com'), 'api.v1');
});

test('get subdomain returns null for no subdomain', () => {
    assert.strictEqual(urlUtils.getSubdomain('https://example.com'), null);
});

test('get subdomain returns null for localhost', () => {
    assert.strictEqual(urlUtils.getSubdomain('https://localhost'), null);
});

// Test isAbsolute() and isRelative()
console.log('\n--- isAbsolute() / isRelative() ---');

test('isAbsolute for full URL', () => {
    assert.strictEqual(urlUtils.isAbsolute('https://example.com'), true);
});

test('isAbsolute for protocol-relative URL', () => {
    assert.strictEqual(urlUtils.isAbsolute('//example.com'), true);
});

test('isAbsolute for relative path returns false', () => {
    assert.strictEqual(urlUtils.isAbsolute('/path'), false);
});

test('isRelative for relative path', () => {
    assert.strictEqual(urlUtils.isRelative('/path'), true);
});

test('isRelative for full URL returns false', () => {
    assert.strictEqual(urlUtils.isRelative('https://example.com'), false);
});

// Test addQuery() and removeQuery()
console.log('\n--- addQuery() / removeQuery() ---');

test('add query parameter', () => {
    const result = urlUtils.addQuery('https://example.com', { key: 'value' });
    assert.strictEqual(result, 'https://example.com/?key=value');
});

test('add query to existing query', () => {
    const result = urlUtils.addQuery('https://example.com?a=1', { b: '2' });
    assert.strictEqual(result, 'https://example.com/?a=1&b=2');
});

test('update existing query parameter', () => {
    const result = urlUtils.addQuery('https://example.com?key=old', { key: 'new' });
    assert.strictEqual(result, 'https://example.com/?key=new');
});

test('remove query parameter', () => {
    const result = urlUtils.removeQuery('https://example.com?a=1&b=2', ['a']);
    assert.strictEqual(result, 'https://example.com/?b=2');
});

test('remove multiple query parameters', () => {
    const result = urlUtils.removeQuery('https://example.com?a=1&b=2&c=3', ['a', 'c']);
    assert.strictEqual(result, 'https://example.com/?b=2');
});

test('addQuery returns null for invalid URL', () => {
    assert.strictEqual(urlUtils.addQuery('not a url', { key: 'value' }), null);
});

// Test getExtension() and getFilename()
console.log('\n--- getExtension() / getFilename() ---');

test('get extension from URL', () => {
    assert.strictEqual(urlUtils.getExtension('https://example.com/file.txt'), 'txt');
});

test('get extension from URL with query', () => {
    assert.strictEqual(urlUtils.getExtension('https://example.com/file.jpg?size=large'), 'jpg');
});

test('get extension returns null for no extension', () => {
    assert.strictEqual(urlUtils.getExtension('https://example.com/path/'), null);
});

test('get filename from URL', () => {
    assert.strictEqual(urlUtils.getFilename('https://example.com/path/to/file.txt'), 'file.txt');
});

test('get filename from URL with query', () => {
    assert.strictEqual(urlUtils.getFilename('https://example.com/document.pdf?v=1'), 'document.pdf');
});

test('get filename returns null for root path', () => {
    assert.strictEqual(urlUtils.getFilename('https://example.com/'), null);
});

// Test toAbsolute()
console.log('\n--- toAbsolute() ---');

test('toAbsolute converts relative to absolute', () => {
    assert.strictEqual(
        urlUtils.toAbsolute('/path', 'https://example.com'),
        'https://example.com/path'
    );
});

test('toAbsolute keeps absolute URL unchanged', () => {
    assert.strictEqual(
        urlUtils.toAbsolute('https://other.com/path', 'https://example.com'),
        'https://other.com/path'
    );
});

// Test areEquivalent()
console.log('\n--- areEquivalent() ---');

test('equivalent URLs', () => {
    assert.strictEqual(
        urlUtils.areEquivalent('https://EXAMPLE.COM/path/', 'https://example.com/path'),
        true
    );
});

test('equivalent URLs with different query order', () => {
    assert.strictEqual(
        urlUtils.areEquivalent('https://example.com?a=1&b=2', 'https://example.com?b=2&a=1'),
        true
    );
});

test('non-equivalent URLs', () => {
    assert.strictEqual(
        urlUtils.areEquivalent('https://example.com/a', 'https://example.com/b'),
        false
    );
});

// Test getPathSegments()
console.log('\n--- getPathSegments() ---');

test('get path segments', () => {
    assert.deepStrictEqual(
        urlUtils.getPathSegments('https://example.com/a/b/c'),
        ['a', 'b', 'c']
    );
});

test('get path segments from root', () => {
    assert.deepStrictEqual(
        urlUtils.getPathSegments('https://example.com/'),
        []
    );
});

test('get path segments handles extra slashes', () => {
    assert.deepStrictEqual(
        urlUtils.getPathSegments('https://example.com//a//b//'),
        ['a', 'b']
    );
});

// Test fromTemplate() and parseTemplate()
console.log('\n--- fromTemplate() / parseTemplate() ---');

test('build URL from template', () => {
    const result = urlUtils.fromTemplate('/users/:id/posts/:postId', { id: '123', postId: '456' });
    assert.strictEqual(result, '/users/123/posts/456');
});

test('build URL from template encodes values', () => {
    const result = urlUtils.fromTemplate('/search/:query', { query: 'hello world' });
    assert.strictEqual(result, '/search/hello%20world');
});

test('parse template parameters', () => {
    const result = urlUtils.parseTemplate('/users/:id/posts/:postId');
    assert.deepStrictEqual(result, ['id', 'postId']);
});

test('parse template with no parameters', () => {
    const result = urlUtils.parseTemplate('/static/path');
    assert.deepStrictEqual(result, []);
});

// Test roundtrip: parse -> build
console.log('\n--- Roundtrip Tests ---');

test('roundtrip: parse then build', () => {
    const original = 'https://user:pass@example.com:8080/path?key=value#hash';
    const parsed = urlUtils.parse(original);
    const built = urlUtils.build(parsed);
    assert.strictEqual(built, original);
});

test('roundtrip: parseQuery then buildQuery', () => {
    const original = 'key1=value1&key2=value2';
    const parsed = urlUtils.parseQuery(original);
    const built = urlUtils.buildQuery(parsed, { prefix: '' });
    assert.strictEqual(built, original);
});

// Print summary
console.log('\n=== Test Summary ===');
console.log(`Passed: ${passCount}`);
console.log(`Failed: ${failCount}`);
console.log(`Total: ${passCount + failCount}`);
console.log(failCount === 0 ? '\n✓ All tests passed!' : `\n✗ ${failCount} test(s) failed.`);

process.exit(failCount === 0 ? 0 : 1);

File: Lua/url_utils/url_utils_test.lua
---
-- URL Utilities Test Suite
-- URL 工具函数测试套件
--
-- @author AllToolkit
-- @version 1.0.0

-- 加载模块
package.path = package.path .. ";./?.lua;./?/init.lua;./mod.lua"
local UrlUtils = dofile("mod.lua")

-- 测试工具函数
local function assertEquals(actual, expected, message)
    if actual ~= expected then
        print("[FAIL] " .. (message or "Assertion failed"))
        print("  Expected: " .. tostring(expected))
        print("  Actual: " .. tostring(actual))
        return false
    end
    return true
end

local function assertTrue(value, message)
    if not value then
        print("[FAIL] " .. (message or "Expected true"))
        return false
    end
    return true
end

local function assertFalse(value, message)
    if value then
        print("[FAIL] " .. (message or "Expected false"))
        return false
    end
    return true
end

local function assertNotNil(value, message)
    if value == nil then
        print("[FAIL] " .. (message or "Expected non-nil value"))
        return false
    end
    return true
end

local function assertNil(value, message)
    if value ~= nil then
        print("[FAIL] " .. (message or "Expected nil"))
        print("  Actual: " .. tostring(value))
        return false
    end
    return true
end

-- 测试计数器
local total_tests = 0
local passed_tests = 0

local function runTest(name, func)
    total_tests = total_tests + 1
    local success, err = pcall(func)
    if success then
        passed_tests = passed_tests + 1
        print("[PASS] " .. name)
    else
        print("[FAIL] " .. name .. ": " .. tostring(err))
    end
end

-- =====================
-- 测试开始
-- =====================

print("\n========================================")
print("URL Utilities Test Suite")
print("========================================\n")

-- =====================
-- URL 编码解码测试
-- =====================

print("--- URL 编码解码测试 ---\n")

runTest("url_encode: 基本编码", function()
    assertEquals(UrlUtils.url_encode("hello world"), "hello%20world")
    assertEquals(UrlUtils.url_encode("a+b=c"), "a%2Bb%3Dc")
    assertEquals(UrlUtils.url_encode("测试"), "%E6%B5%8B%E8%AF%95")
end)

runTest("url_encode: 空值处理", function()
    assertNil(UrlUtils.url_encode(nil))
    assertEquals(UrlUtils.url_encode(""), "")
end)

runTest("url_decode: 基本解码", function()
    assertEquals(UrlUtils.url_decode("hello%20world"), "hello world")
    assertEquals(UrlUtils.url_decode("a%2Bb%3Dc"), "a+b=c")
    assertEquals(UrlUtils.url_decode("%E6%B5%8B%E8%AF%95"), "测试")
end)

runTest("url_decode: 空值处理", function()
    assertNil(UrlUtils.url_decode(nil))
    assertEquals(UrlUtils.url_decode(""), "")
end)

runTest("encode_query: 查询编码", function()
    assertEquals(UrlUtils.encode_query("hello world"), "hello+world")
    assertEquals(UrlUtils.encode_query("a&b=c"), "a%26b%3Dc")
end)

runTest("decode_query: 查询解码", function()
    assertEquals(UrlUtils.decode_query("hello+world"), "hello world")
    assertEquals(UrlUtils.decode_query("a%26b%3Dc"), "a&b=c")
end)

runTest("encode_path: 路径编码", function()
    assertEquals(UrlUtils.encode_path("/path/to/file"), "/path/to/file")
    assertEquals(UrlUtils.encode_path("/path with space"), "/path%20with%20space")
end)

-- =====================
-- 查询字符串测试
-- =====================

print("\n--- 查询字符串测试 ---\n")

runTest("parse_query: 基本解析", function()
    local params = UrlUtils.parse_query("name=John&age=30")
    assertEquals(params.name, "John")
    assertEquals(params.age, "30")
end)

runTest("parse_query: URL 编码", function()
    local params = UrlUtils.parse_query("name=John%20Doe&city=New%20York")
    assertEquals(params.name, "John Doe")
    assertEquals(params.city, "New York")
end)

runTest("parse_query: 空值", function()
    local params = UrlUtils.parse_query("name=&age=30")
    assertEquals(params.name, "")
    assertEquals(params.age, "30")
end)

runTest("parse_query: 带 ? 前缀", function()
    local params = UrlUtils.parse_query("?name=John")
    assertEquals(params.name, "John")
end)

runTest("parse_query: 空字符串", function()
    local params = UrlUtils.parse_query("")
    assertEquals(next(params), nil)
end)

runTest("parse_query: nil 处理", function()
    local params = UrlUtils.parse_query(nil)
    assertEquals(next(params), nil)
end)

runTest("build_query: 基本构建", function()
    local query = UrlUtils.build_query({ name = "John", age = "30" })
    assertTrue(query:find("name=John") ~= nil)
    assertTrue(query:find("age=30") ~= nil)
end)

runTest("build_query: URL 编码", function()
    local query = UrlUtils.build_query({ name = "John Doe" })
    assertEquals(query, "name=John+Doe")
end)

runTest("build_query: 带 ? 前缀", function()
    local query = UrlUtils.build_query({ name = "John" }, true)
    assertEquals(query, "?name=John")
end)

runTest("merge_query: 合并参数", function()
    local url = UrlUtils.merge_query("http://example.com/path?a=1", { b = "2", c = "3" })
    assertTrue(url:find("a=1") ~= nil)
    assertTrue(url:find("b=2") ~= nil)
    assertTrue(url:find("c=3") ~= nil)
end)

-- =====================
-- URL 解析测试
-- =====================

print("\n--- URL 解析测试 ---\n")

runTest("parse: 完整 URL", function()
    local parsed = UrlUtils.parse("https://user:pass@example.com:8080/path?query=value#fragment")
    assertEquals(parsed.scheme, "https")
    assertEquals(parsed.userinfo, "user:pass")
    assertEquals(parsed.host, "example.com")
    assertEquals(parsed.port, 8080)
    assertEquals(parsed.path, "/path")
    assertEquals(parsed.query, "query=value")
    assertEquals(parsed.fragment, "fragment")
end)

runTest("parse: 简单 URL", function()
    local parsed = UrlUtils.parse("http://example.com")
    assertEquals(parsed.scheme, "http")
    assertEquals(parsed.host, "example.com")
    assertEquals(parsed.path, "/")
    assertNil(parsed.port)
    assertNil(parsed.query)
    assertNil(parsed.fragment)
end)

runTest("parse: 带路径", function()
    local parsed = UrlUtils.parse("http://example.com/path/to/resource")
    assertEquals(parsed.host, "example.com")
    assertEquals(parsed.path, "/path/to/resource")
end)

runTest("parse: 带端口", function()
    local parsed = UrlUtils.parse("http://example.com:8080")
    assertEquals(parsed.host, "example.com")
    assertEquals(parsed.port, 8080)
end)

runTest("parse: 带 IPv6 地址", function()
    local parsed = UrlUtils.parse("http://[::1]:8080/path")
    assertEquals(parsed.host, "[::1]")
    assertEquals(parsed.port, 8080)
end)

runTest("parse: nil 处理", function()
    assertNil(UrlUtils.parse(nil))
    assertNil(UrlUtils.parse(""))
end)

runTest("build: 完整 URL", function()
    local url = UrlUtils.build({
        scheme = "https",
        host = "example.com",
        port = 443,
        path = "/path",
        query = "a=1",
        fragment = "section",
    })
    assertTrue(url:find("^https://example.com") ~= nil)
    assertTrue(url:find("/path") ~= nil)
    assertTrue(url:find("%?a=1") ~= nil)
    assertTrue(url:find("#section") ~= nil)
end)

runTest("build: 最小 URL", function()
    local url = UrlUtils.build({ host = "example.com" })
    assertTrue(url:find("example.com") ~= nil)
end)

-- =====================
-- URL 验证测试
-- =====================

print("\n--- URL 验证测试 ---\n")

runTest("is_valid: 有效 URL", function()
    assertTrue(UrlUtils.is_valid("http://example.com"))
    assertTrue(UrlUtils.is_valid("https://example.com/path"))
    assertTrue(UrlUtils.is_valid("ftp://ftp.example.com"))
end)

runTest("is_valid: 无效 URL", function()
    assertFalse(UrlUtils.is_valid(nil))
    assertFalse(UrlUtils.is_valid(""))
    assertFalse(UrlUtils.is_valid("not a url"))
end)

runTest("is_http: HTTP URL", function()
    assertTrue(UrlUtils.is_http("http://example.com"))
    assertTrue(UrlUtils.is_http("https://example.com"))
    assertFalse(UrlUtils.is_http("ftp://example.com"))
end)

runTest("is_safe: 安全 URL", function()
    assertTrue(UrlUtils.is_safe("https://example.com"))
    assertFalse(UrlUtils.is_safe("https://user:pass@example.com"))
    assertFalse(UrlUtils.is_safe("ftp://example.com"))
end)

runTest("default_port: 默认端口", function()
    assertEquals(UrlUtils.default_port("http"), 80)
    assertEquals(UrlUtils.default_port("https"), 443)
    assertEquals(UrlUtils.default_port("ftp"), 21)
    assertNil(UrlUtils.default_port("unknown"))
end)

-- =====================
-- URL 规范化测试
-- =====================

print("\n--- URL 规范化测试 ---\n")

runTest("normalize: 大小写规范化", function()
    local url = UrlUtils.normalize("HTTP://EXAMPLE.COM/Path")
    assertTrue(url ~= nil and url:lower():find("http://example.com") ~= nil)
end)

runTest("normalize: 移除默认端口", function()
    local url = UrlUtils.normalize("http://example.com:80/path")
    assertTrue(url ~= nil and url:find(":80") == nil)
end)

runTest("normalize_path: 路径规范化", function()
    assertEquals(UrlUtils.normalize_path("/a/./b/../c/"), "/a/c/")
    assertEquals(UrlUtils.normalize_path("/a/b/../../c"), "/c")
end)

-- =====================
-- URL 操作测试
-- =====================

print("\n--- URL 操作测试 ---\n")

runTest("get_extension: 获取扩展名", function()
    assertEquals(UrlUtils.get_extension("http://example.com/file.txt"), "txt")
    assertEquals(UrlUtils.get_extension("http://example.com/path/image.png"), "png")
    assertNil(UrlUtils.get_extension("http://example.com/path/"))
end)

runTest("get_filename: 获取文件名", function()
    assertEquals(UrlUtils.get_filename("http://example.com/path/file.txt"), "file.txt")
    assertEquals(UrlUtils.get_filename("http://example.com/document.pdf"), "document.pdf")
end)

runTest("get_domain: 获取域名", function()
    assertEquals(UrlUtils.get_domain("http://www.example.com/path"), "www.example.com")
    assertEquals(UrlUtils.get_domain("https://sub.domain.example.com:8080"), "sub.domain.example.com")
end)

runTest("get_tld: 获取顶级域名", function()
    assertEquals(UrlUtils.get_tld("http://www.example.com"), "example.com")
    assertEquals(UrlUtils.get_tld("http://test.co.uk"), "co.uk")
end)

runTest("get_subdomain: 获取子域名", function()
    assertEquals(UrlUtils.get_subdomain("http://www.example.com"), "www")
    assertEquals(UrlUtils.get_subdomain("http://a.b.c.example.com"), "a.b.c")
    assertNil(UrlUtils.get_subdomain("http://example.com"))
end)

runTest("join: 路径拼接", function()
    local url = UrlUtils.join("http://example.com/path", "subdir")
    assertTrue(url ~= nil and url:find("/path/subdir") ~= nil)
end)

runTest("join: 绝对路径", function()
    local url = UrlUtils.join("http://example.com/path", "/newpath")
    assertTrue(url ~= nil and url:find("/newpath") ~= nil)
end)

runTest("resolve: 相对 URL 解析", function()
    local url = UrlUtils.resolve("/path", "http://example.com/base")
    assertEquals(url, "http://example.com/path")
end)

runTest("is_same_origin: 同源判断", function()
    assertTrue(UrlUtils.is_same_origin(
        "http://example.com/a",
        "http://example.com/b"
    ))
    assertFalse(UrlUtils.is_same_origin(
        "http://example.com/a",
        "http://other.com/b"
    ))
end)

-- =====================
-- URL 参数提取测试
-- =====================

print("\n--- URL 参数提取测试 ---\n")

runTest("get_param: 获取参数", function()
    assertEquals(UrlUtils.get_param("http://example.com?name=John&age=30", "name"), "John")
    assertEquals(UrlUtils.get_param("http://example.com?name=John&age=30", "age"), "30")
    assertNil(UrlUtils.get_param("http://example.com?name=John", "missing"))
end)

runTest("set_param: 设置参数", function()
    local url = UrlUtils.set_param("http://example.com?a=1", "b", "2")
    assertTrue(url ~= nil and url:find("b=2") ~= nil)
end)

runTest("remove_param: 删除参数", function()
    local url = UrlUtils.remove_param("http://example.com?a=1&b=2", "a")
    assertTrue(url ~= nil and url:find("a=") == nil)
    assertTrue(url ~= nil and url:find("b=2") ~= nil)
end)

-- =====================
-- 实用工具测试
-- =====================

print("\n--- 实用工具测试 ---\n")

runTest("hash: URL 哈希", function()
    local hash1 = UrlUtils.hash("http://example.com/path")
    local hash2 = UrlUtils.hash("http://example.com/path")
    local hash3 = UrlUtils.hash("http://example.com/other")
    
    assertEquals(hash1, hash2)
    assertTrue(hash1 ~= hash3)
    assertEquals(#hash1, 8)
end)

runTest("builder: URL 构建器", function()
    local url = UrlUtils.builder("https", "example.com")
        :set_port(8080)
        :set_path("/api/users")
        :add_param("page", "1")
        :add_param("limit", "10")
        :set_fragment("results")
        :build()
    
    assertTrue(url ~= nil and url:find("https://example.com:8080") ~= nil)
    assertTrue(url ~= nil and url:find("/api/users") ~= nil)
    assertTrue(url ~= nil and url:find("page=1") ~= nil)
    assertTrue(url ~= nil and url:find("#results") ~= nil)
end)

-- =====================
-- 测试结果统计
-- =====================

print("\n========================================")
print(string.format("测试结果: %d/%d 通过", passed_tests, total_tests))
print("========================================\n")

if passed_tests == total_tests then
    print("✓ 所有测试通过！")
    os.exit(0)
else
    print("✗ 存在失败的测试")
    os.exit(1)
end